Source/Details/ASCollectionViewLayout.m

```diff
@@ -0,0 +1,162 @@
+
+- (void)prepareLayout
+{
+  if (_collectionViewIsPendingLayoutUpdate == NO && _elementMap != nil) {
+    ASLayout *layout = [_calculator layoutThatFits:self.collectionView.bounds.size forElementMap:_elementMap];
+    [self _applyLayout:layout];
+  }
+}
+
+- (void)invalidateLayout
+{
+  [super invalidateLayout];
+  if (_collectionViewIsPendingLayoutUpdate == NO) {
+    _contentSize = CGSizeZero;
+    [_elementToLayoutAttributesMap removeAllObjects];
+  }
+}
+
+- (CGSize)collectionViewContentSize
+{
+  _collectionViewIsPendingLayoutUpdate = NO;
+  return _contentSize;
+}
+
+- (NSArray<UICollectionViewLayoutAttributes *> *)layoutAttributesForElementsInRect:(CGRect)rect
+{
+  NSMutableArray *attributesInRect = [NSMutableArray array];
+  NSEnumerator<UICollectionViewLayoutAttributes *> *enumerator = [_elementToLayoutAttributesMap objectEnumerator];
+  UICollectionViewLayoutAttributes *attrs;
+  while ((attrs = [enumerator nextObject])) {
+    if (CGRectIntersectsRect(rect, attrs.frame)) {
+      [attributesInRect addObject:attrs];
+    }
+  }
+  return attributesInRect;
+}
+
+- (UICollectionViewLayoutAttributes *)layoutAttributesForItemAtIndexPath:(NSIndexPath *)indexPath
+{
+  ASCollectionElement *element = [_elementMap elementForItemAtIndexPath:indexPath];
+  return [_elementToLayoutAttributesMap objectForKey:element];
+}
+
+- (UICollectionViewLayoutAttributes *)layoutAttributesForSupplementaryViewOfKind:(NSString *)elementKind atIndexPath:(NSIndexPath *)indexPath
+{
+  ASCollectionElement *element = [_elementMap supplementaryElementOfKind:elementKind atIndexPath:indexPath];
+  return [_elementToLayoutAttributesMap objectForKey:element];
+}
+
+- (UICollectionViewLayoutAttributes *)layoutAttributesForDecorationViewOfKind:(NSString *)elementKind atIndexPath:(NSIndexPath *)indexPath
+{
+  return [self layoutAttributesForSupplementaryViewOfKind:elementKind atIndexPath:indexPath];
+}
+
+- (BOOL)shouldInvalidateLayoutForBoundsChange:(CGRect)newBounds
+{
+  if (!CGSizeEqualToSize(self.collectionView.bounds.size, newBounds.size)) {
+    return YES;
+  }
+  return NO;
+}
+
+#pragma mark - Private methods
+
+- (void)_applyLayout:(ASLayout *)layout
+{
+  ASDisplayNodeAssertMainThread();
+  _contentSize = layout.size;
+  
+  for (ASLayout *sublayout in layout.sublayouts) {
+    ASCollectionElement *element = ((ASCellNode *)sublayout.layoutElement).collectionElement;
+    NSIndexPath *indexPath = [_elementMap indexPathForElement:element];
+    
+    UICollectionViewLayoutAttributes *attrs;
+    if (element.supplementaryElementKind == nil) {
+      attrs = [UICollectionViewLayoutAttributes layoutAttributesForCellWithIndexPath:indexPath];
+    } else {
+      attrs = [UICollectionViewLayoutAttributes layoutAttributesForSupplementaryViewOfKind:element.supplementaryElementKind withIndexPath:indexPath];
+    }
+    
+    attrs.frame = CGRectMake(sublayout.position.x, sublayout.position.y, sublayout.size.width, sublayout.size.height);
```

Comment:
For both readability and performance, suggest using a CGPoint and CGSize local variable.